base/strings/annotated.jl

```diff
@@ -386,3 +386,94 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+end
+
+AnnotatedIOBuffer(io::IOBuffer) = AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, aio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', aio.io.size, " byte", ifelse(aio.io.size == 1, "", "s"), ", ",
+          length(aio.annotations), " annotation", ifelse(length(aio.annotations) == 1, "", "s"), ")")
+end
+
+isopen(io::AnnotatedIOBuffer) = isopen(io.io)
+close(io::AnnotatedIOBuffer) = close(io.io)
+closewrite(io::AnnotatedIOBuffer) = closewrite(io.io)
+eof(io::AnnotatedIOBuffer) = eof(io.io)
+peek(io::AnnotatedIOBuffer) = peek(io.io)
+position(io::AnnotatedIOBuffer) = position(io.io)
+seek(io::AnnotatedIOBuffer, n::Integer) = (seek(io.io, n); io)
+skip(io::AnnotatedIOBuffer, n::Integer) = (skip(io.io, n); io)
+mark(io::AnnotatedIOBuffer) = mark(io.io)
+reset(io::AnnotatedIOBuffer) = reset(io.io)
+unmark(io::AnnotatedIOBuffer) = unmark(io.io)
+ismarked(io::AnnotatedIOBuffer) = ismarked(io.io)
+copy(io::AnnotatedIOBuffer) = AnnotatedIOBuffer(copy(io.io), copy(io.annotations))
+unsafe_write(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_write(io.io, b, len)
+unsafe_read(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_read(io.io, b, len)
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    for (region, annot) in astr.annotations
+        start, stop = first(region), last(region)
+        push!(io.annotations, (start+offset:stop+offset, annot))
+    end
+    write(io.io, String(astr))
```

Comment:
Should this also first erase any existing annotations for this region in the `io`?

Reply:
Probably I think. This is what an implementation of that looks like:

```julia
if !eof(io)
    # If we are overwriting an existing span in the AnnotatedIOBuffer,
    # clear out any overlapping pre-existing annotations.
    span = offset:offset+ncodeunits(astr)
    filter!(((range, _),) -> first(span) <= first(range) && last(range) <= last(span), io.annotations)
    extras = Tuple{UnitRange{Int}, Pair{Symbol, Any}}[]
    for i in eachindex(io.annotations)
        range, annot = io.annotations[i]
        # Test for partial overlap
        if first(span) <= first(range) <= last(span) || first(span) <= last(range) <= last(span)
            io.annotations[i] = (if first(range) < first(span)
                                     first(range):first(span)-1
                                 else last(span)+1:last(range) end, annot)
            # If `span` fits exactly within `range`, then we've only copied over
            # the beginning overhang, but also need to conserve the end overhang.
            if first(range) < first(span) && last(span) < last(range)
                io.annotations[i] = (first(range):first(span)-1, annot)
                push!(extras, (last(span)+1:last(range), annot))
            end
        end
        # Insert any extra entries in the appropriate position
        for entry in extras
            indices = searchsorted(io.annotations, (first(entry),), by=first)
            splice!(s.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
        end
    end
end
```

Reply:
Implemented with tests.